internal/controller/genericprovider_controller.go

```diff
@@ -106,7 +112,35 @@ func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile
 		return r.reconcileDelete(ctx, typedProvider)
 	}
 
-	return r.reconcile(ctx, typedProvider, typedProviderList)
+	// Check if spec hash stays the same and don't go further in this case.
+	specHash, err := calculateHash(typedProvider.GetSpec())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if typedProvider.GetAnnotations()[appliedSpecHashAnnotation] == specHash {
```

Comment:
If I remember it right, `GetAnnotations()` can still return nil

Reply:
Yeah, it can, but it doesn't matter. We can get values from nil maps. Here is an example I wrote: https://go.dev/play/p/pUL9u7U5OyR